File: sdk/python/examples/cached_client.py
"""Cached client: opens one stream, keeps a snapshot in memory, and serves
synchronous local reads with zero network per read. Use this for a long-lived
process. Run with:  FLAG_IT_SDK_KEY=sdk-… python cached_client.py
"""

import os
import signal

import flagit

client = flagit.CachedClient(
    base_url=os.environ.get("FLAG_IT_URL", "http://localhost:8080"),
    sdk_key=os.environ["FLAG_IT_SDK_KEY"],
    context=flagit.context("user", "u-123", {"plan": "pro"}),
)

# Blocks until the first snapshot has streamed in.
client.wait_for_initialization(timeout=5)
print("initialized:", client.is_initialized())


def on_change() -> None:
    # Fired when a change is pushed over the stream — no reload, no polling.
    print("flags changed → new-dashboard:", client.variation("new-dashboard", False))


unsubscribe = client.on_change(on_change)

# Synchronous local reads — no network:
print("new-dashboard:", client.variation("new-dashboard", False))
print("tier:", client.string_variation("pricing-tier", "free"))  # typed convenience

# Block, reacting to live changes; Ctrl-C to stop.
try:
    signal.pause()
except KeyboardInterrupt:
    pass
finally:
    unsubscribe()
    client.close()  # stops the stream and flushes a final batch of usage events

File: sdk/javascript/examples/cached-client.mjs
// Cached client: opens one stream, keeps a snapshot in memory, and serves
// synchronous local reads with zero network per read. Use this for anything
// long-lived and read-heavy. Run with:
//   FLAG_IT_SDK_KEY=sdk-… node cached-client.mjs
import { createCachedClient, context } from '@flag-it/sdk'

const client = createCachedClient({
  baseUrl: process.env.FLAG_IT_URL ?? 'http://localhost:8080',
  sdkKey: process.env.FLAG_IT_SDK_KEY,
  context: context('user', 'u-123', { plan: 'pro' }),
})

// Resolves once the first snapshot has streamed in.
await client.waitForInitialization()

// React to live changes pushed over the stream (admin flips a flag, a rollout
// widens, targeting changes) — no reload, no polling.
const unsubscribe = client.onChange(() => {
  console.log('flags changed → new-dashboard:', client.boolVariation('new-dashboard', false))
})

// Synchronous local reads — no await needed for the value, no network:
console.log('new-dashboard:', client.boolVariation('new-dashboard', false))
console.log('tier:', client.stringVariation('pricing-tier', 'free'))

// On logout / account switch, re-point at the new identity:
await client.identify(context('user', 'u-456'))
console.log('after identify → new-dashboard:', client.boolVariation('new-dashboard', false))

// Graceful shutdown flushes a final batch of usage events.
process.on('SIGINT', async () => {
  unsubscribe()
  await client.close()
  process.exit(0)
})
